feat/segmentation mode

sedghi opened the pull request:
### Context

Adds new segmentation mode with brushes, add new segmentation and export capabilities

~~Requires this PR to be merged first https://github.com/cornerstonejs/cornerstone3D/pull/771~~

Fixes various issues such as https://github.com/OHIF/Viewers/issues/3660 

## 🎉  Kudos @dan-rukas for designing the UI elements

### Changes & Results

#### Add new segmentation on viewports and edit via Segmentation tools and export


https://github.com/OHIF/Viewers/assets/7490180/f644582c-6ba0-470f-bc30-75ab298b0e1d




####  toolGroupService.createToolGroupAndAddTools

this function has changed to make it simpler, now the tool hosts the configuration too

Before

```js
const tools = {
  passive: [
    {
      toolName: toolNames.ArrowAnnotate,
    },
  ],
  // the rest of tools
}

const configuration = {
  // the rest of configuration
  [toolNames.ArrowAnnotate]: {
    getTextCallback: (callback, eventDetails) => {
      commandsManager.runCommand("arrowTextCallback", {
        callback,
        eventDetails,
      })
    },

    changeTextCallback: (data, eventDetails, callback) =>
      commandsManager.runCommand("arrowTextCallback", {
        callback,
        data,
        eventDetails,
      }),
  },
}

toolGroupService.createToolGroupAndAddTools("toolGroupId", tools, configuration)

```

After 

```js
const tools = {
  // the rest
  passive: [
    {
      toolName: toolNames.ArrowAnnotate,
      configuration: {
        getTextCallback: (callback, eventDetails) => {
          commandsManager.runCommand("arrowTextCallback", {
            callback,
            eventDetails,
          })
        },

        changeTextCallback: (data, eventDetails, callback) =>
          commandsManager.runCommand("arrowTextCallback", {
            callback,
            data,
            eventDetails,
          }),
      },
    },
  ],
  // the rest of tools
}

toolGroupService.createToolGroupAndAddTools('toolGroupId', tools);

```



### Testing

- Segment on https://deploy-preview-3632--ohif-dev.netlify.app/segmentation?StudyInstanceUIDs=2.16.840.1.114362.1.11972228.22789312658.616067305.306.2
- For export (create report) you should run it with your local server (e.g., orthanc), however, you should be able to Download the segmentation 

### Checklist

#### PR

<!--
https://semantic-release.gitbook.io/semantic-release/#how-does-it-work

Examples:
Please note the letter casing in the provided examples (upper or lower).

- feat(MeasurementService): add ...
- fix(Toolbar): fix ...
- docs(Readme): update ...
- style(Whitespace): fix ...
- refactor(ExtensionManager): ...
- test(HangingProtocol): Add test ...
- chore(git): update ...
- perf(VolumeLoader): ...

You don't need to have each commit within the Pull Request follow the rule,
but the PR title must comply with it, as it will be used as the commit message
after the commits are squashed.
-->

- [] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

<!-- https://docs.ohif.org/ -->

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [] OS: <!--[e.g. Windows 10, macOS 10.15.4]-->
- [] Node version: <!--[e.g. 16.14.0]-->
- [] Browser:
  <!--[e.g. Chrome 83.0.4103.116, Firefox 77.0.1, Safari 13.1.1]-->

<!-- prettier-ignore-start -->
[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch
<!-- prettier-ignore-end -->

SachidanandAlle commented:
Thank for getting this out as a basic feature in ohif viewer 👍

jbocce commented:
I have not started looking at the code changes yet, but I have started to briefly test the new segmentation mode.

It is awesome to see that the segmentations added in non-mpr, display in mpr. Furthermore, the segmentation tools appear to work very well in mpr as well! Good stuff!

I have found the following issues thus far:

1. There is a noticeable delay when clicking Add Segmentation. About 1-2 seconds (it does vary - and appears longer when running locally against orthanc). Is this expected? If feels a bit long to me.
2. If one selects any brush before creating a segmentation and then clicks and drags on an image, nothing occurs in the UI to suggest a segmentation must be created first. Indeed a warning message is output in the console, but I suspect most users will not have the console open. Perhaps the brush tools can be disabled until a segmentation is created? Note also that the following exception is also thrown in the console...
   ```
   BrushTool.ts:112 Uncaught Error: Not implemented yet
    at Rr.preMouseDownCallback (BrushTool.ts:112:13)
    at HTMLDivElement.fg (mouseDown.ts:60:38)
    at se (triggerEvent.ts:37:13)
    at mouseDownListener.ts:214:23
    at mt (mouseDownListener.ts:406:3)
    at HTMLDocument.dt (mouseDownListener.ts:245:7)
   ```
3. Creating a subsequent segmentation (i.e. via the Add New Segmentation menu item) does not appear to work.

4. Deleting the first (and only) segmentation throws an exception in the console and the segmentation remains in the side panel. However the segmentation (annotations) on the images are removed.
   ```
   Uncaught TypeError: Cannot set properties of undefined (setting 'isHydrated')
    at ht._setDisplaySetIsHydrated (SegmentationService.ts:1148:16)
    at ht.remove (SegmentationService.ts:1313:10)
    at onSegmentationDelete (PanelSegmentation.tsx:65:25)
    at onClick (SegmentationDropDownRow.tsx:67:17)
    at onClick (Dropdown.tsx:53:11)
    at Object.je (react-dom.production.min.js:52:317)
    at Qe (react-dom.production.min.js:52:471)
    at react-dom.production.min.js:53:35
    at Pr (react-dom.production.min.js:100:68)
    at Or (react-dom.production.min.js:101:380)
   ```	
5. There is some weird rounding that occurs when changing the outline and fill opacity for the segmentation.

![ScreenHunter Sep  08 09 23](https://github.com/OHIF/Viewers/assets/109477394/f1ef3333-2895-41e7-90b5-85627bd18895)

6. When using the sphere brush tool immediately after using the sphere ROI threshold tool, the annotation added by the brush tool does not "act" like a sphere - that is to say navigating forward and backward to neighbouring slices do NOT show the segmented (sphere) annotation as expected. Perhaps the brush tool shows that sphere is selected, but it is really circle that is selected?

https://github.com/OHIF/Viewers/assets/109477394/214b3aa1-b1e9-4f89-bfb8-c54accc50afd

7. I have occassionally (but rarely) seen tooltips for the various segmentation tool (icons). The tooltips should be shown on hover consistently in my opinion to give a better idea of what each tool is/does.

8. When adjusting the radius of any of the circle or sphere tools using the slider, fractions of a mm are displayed. However, I am not able to type in a fraction. Is this expected?

9. I noticed varying delays when first dragging the sphere brush tool. For example, doing a quick initial drag from one point to another draws the initial sphere, but then there is a delay to draw the remainder of the "path" to the second point.
Furthermore, the are occassions where some of the path between the two points is "skipped" (i.e. not filled in).

https://github.com/OHIF/Viewers/assets/109477394/aed014e5-1240-4933-98a6-5573b11d0fda

10. Download segmentation throws an exception and nothing is downloaded.
   ```
   Uncaught TypeError: U is not a function
    at downloadSegmentation (commandsModule.ts:181:7)
    at jt.runCommand (CommandsManager.ts:162:14)
    at onSegmentationDownload (PanelSegmentation.tsx:186:21)
    at onClick (SegmentationDropDownRow.tsx:81:23)
    at onClick (Dropdown.tsx:53:11)
    at Object.je (react-dom.production.min.js:52:317)
    at Qe (react-dom.production.min.js:52:471)
    at react-dom.production.min.js:53:35
    at Pr (react-dom.production.min.js:100:68)
    at Or (react-dom.production.min.js:101:380)
   ```	
11. The lock icon for a segment appears strangely.
![ScreenHunter Sep  08 09 52](https://github.com/OHIF/Viewers/assets/109477394/333e6452-965d-425d-aecc-5e695d13503e)

12. It is not obvious what clicking the lock icon is supposed to do.

13. Delete all segments (but leave the segmentation) and notice that the tools are still available and will draw annotations on the images. I would expect the tools to not even be selectable.

14. Should the measurement tools be available in this new mode? If they are not, then shouldn't we remove them from the top toolbar?

15. When switching back to a tool in the top toolbar, both the last active segmentation tool and the recently selected tool in the top tool bar show as active/selected.

![two tools show as active](https://github.com/OHIF/Viewers/assets/109477394/5c570148-d888-4b46-8c3a-42fd59bc0d39)


16. Add a sphere segment in non-mpr. Change the opacity to something more opaque. Enter and then exit mpr, the opacity now appears to be non-existent (i.e. transparent). Furthermore, adjusting the opacity for the segmentation has NO effect on the segment at all.

17. I was able to Create a Report (i.e. export the segmentation), however when attempting to hydrate the segmentation I see a gray box overlaying the images.


https://github.com/OHIF/Viewers/assets/109477394/ef3591f7-9aeb-4e71-a29d-805aed695941


18. It appears that the segmentation tools can only be applied to the viewport that was active when the segmentation was created. Attempting to apply the tools to another viewport throws the following exception.
   ```
   C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\tools\segmentation\BrushTool.ts:112 Uncaught Error: Not implemented yet
    at BrushTool.preMouseDownCallback (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\tools\segmentation\BrushTool.ts:112:1)
    at HTMLDivElement.mouseDown (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\eventDispatchers\mouseEventHandlers\mouseDown.ts:60:1)
    at triggerEvent (C:\dev\Viewers\node_modules\@cornerstonejs\core\src\utilities\triggerEvent.ts:37:1)
    at _doMouseDown (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\eventListeners\mouse\mouseDownListener.ts:214:1)
    at _doStateMouseDownAndUp (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\eventListeners\mouse\mouseDownListener.ts:406:1)
    at HTMLDocument._onMouseDrag (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\eventListeners\mouse\mouseDownListener.ts:245:1)
   ```	

https://github.com/OHIF/Viewers/assets/109477394/5d1abe4e-8f23-4bf5-adfd-2c8605b50c1b

19. In some cases, creating a segmentation, changes the window level of the active viewport.

https://deploy-preview-3632--ohif-dev.netlify.app/segmentation?StudyInstanceUIDs=1.3.6.1.4.1.14519.5.2.1.3671.4754.298665348758363466150039312520


https://github.com/OHIF/Viewers/assets/109477394/01f35cd7-a477-400b-8fcb-607b438db7c0

sedghi commented:
Thanks for the thorough review 

Item 1 -> The initial delay is kind of inevitable right now since a volume need to be created from the stack viewport (similar to mpr)
Item 9 -> I don't see such delay or mis-behavior, can you try again?
Item 12 -> what do you propose? 
Item 15 -> I can't reproduce it anymore, maybe I fixed it in items 1-14 
Item 17 -> Can't reproduce it anymore
Item 19 -> it doesn't, it just moved the slice to the middle, i think i fixed that too

jbocce commented on extensions/cornerstone/src/Viewport/Overlays/CustomizableViewportOverlay.tsx:
```diff
@@ -254,7 +254,7 @@ function CustomizableViewportOverlay({
       };
 
-      if (item.customizationType === 'ohif.overlayItem.windowLevel') {
+      if (item.customizationType === 'alireza') {
```
Revert this?

jbocce commented on extensions/cornerstone/src/tools/ImageOverlayViewerTool.tsx:
```diff
@@ -46,6 +47,15 @@ class ImageOverlayViewerTool extends AnnotationDisplayTool {
     this._cachedOverlayMetadata = new Map();
   };
 
+  protected getReferencedImageId(viewport: IStackViewport | IVolumeViewport): string {
```
Why can't we use the one from AnnotationDisplayTool?

sedghi replied:
Since that one requires worldPos and other coordinates, which we don't have here

jbocce commented on extensions/cornerstone/src/services/ViewportService/CornerstoneViewportService.ts:
```diff
@@ -589,7 +585,20 @@ class CornerstoneViewportService extends PubSubService implements IViewportServi
+      let segFrameOfReferenceUID = this._getFrameOfReferenceUID(segDisplaySetInstanceUID);
+
+      if (!segFrameOfReferenceUID) {
+        // if the segmentation does not have a FOR, we might check the
```
It is confusing that the word segmentation appears twice in this sentence. Should this say instead, "if the segmentation display set does not..."?

jbocce replied:
Also using FOR is confusing, I would stick with 'frame of reference uid' in the comment.

jbocce commented on platform/ui/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -28,22 +60,31 @@ const Dropdown = ({ id, children, showDropdownIcon, list, titleClassName }) => {
+        </div>
       </div>
     );
   }, []);
```
I think the `useCallback` dependency should included (at least) `maxCharactersPerLine`, but I suppose if it does not change then it might not matter.

jbocce commented on platform/ui/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -118,9 +161,10 @@ const Dropdown = ({ id, children, showDropdownIcon, list, titleClassName }) => {
-
 Dropdown.defaultProps = {
   showDropdownIcon: true,
+  maxTextLength: 10,
```
Should this be `maxCharactersPerLine` instead?

jbocce commented on extensions/cornerstone/src/services/SegmentationService/SegmentationService.ts:
```diff
@@ -281,15 +292,19 @@ class SegmentationService extends PubSubService {
     this._setSegmentLocked(segmentationId, segmentIndex, isLocked, suppressEvents);
   }
 
+  public toggleSegmentLocked(segmentationId: string, segmentIndex: number): void {
```
Add to docs?

jbocce commented on extensions/cornerstone/src/utils/stackSync/toggleStackImageSync.ts:
```diff
@@ -2,7 +2,7 @@ import calculateViewportRegistrations from './calculateViewportRegistrations';
 // [ {
 //   synchronizerId: string,
-//   viewports: [ { viewportId: string, renderingEngineId: string, index: number } , ...]
+//   viewports: [ { viewportId: number, renderingEngineId: string, index: number } , ...]
```
I realize this is in a comment, but I think `viewportId` is a string.

jbocce commented on extensions/cornerstone-dicom-seg/src/panels/Toolbox/ThresholdSettingsPreset.tsx:
```diff
@@ -0,0 +1,88 @@
+
+    onRangeChange(newRange);
+
+    const updatedOptions = options.map(o => {
```
Hmmm... I believe it would be safest to do this as part of the update function that can be passed to `setOptions`?

```
setOptions((options) => {
  return options.map(...)
});
```

sedghi replied:
I forgot to remove this file, it is outdated and we don't need it

jbocce commented on extensions/cornerstone-dicom-seg/src/panels/Toolbox/ThresholdSettingsPreset.tsx:
```diff
@@ -0,0 +1,88 @@
+      <div className="mt-1 w-1/2">
+        <Select
+          isClearable={false}
+          onChange={handlePresetChange}
```
Where is `handlePresetChange`?

jbocce commented on extensions/cornerstone-dicom-seg/src/panels/Toolbox/ThresholdSettingsPreset.tsx:
```diff
@@ -0,0 +1,88 @@
+  },
+];
+
+function ThresholdSettings({ onRangeChange }) {
```
Is this referenced somewhere?

jbocce commented on extensions/cornerstone-dicom-seg/src/panels/SegmentationToolbox.tsx:
```diff
@@ -0,0 +1,372 @@
+const { segmentation: segmentationUtils } = utilities;
+
+const ACTIONS = {
+  SET_BRUSH_SIZE: 'SET_BRUSH_SIZE',
```
Is this needed anymore? I don't think it is being used.

jbocce commented on platform/ui/src/components/AdvancedToolbox/AdvancedToolbox.tsx:
```diff
@@ -0,0 +1,64 @@
+import ToolSettings from './ToolSettings';
+
+const AdvancedToolbox = ({ title, items }) => {
+  const [isActive, setIsActive] = useState(null);
```
`isActive` sounds like a boolean to me, yet I believe this state is for the name of the active item. So consider using `[activeItemName, setActiveItemName]` instead.

jbocce commented on platform/ui/src/components/AdvancedToolbox/ToolSettings.tsx:
```diff
@@ -0,0 +1,78 @@
+  return (
+    <div className="space-y-2 py-2 text-white">
+      {options?.map(option => {
+        if (option.type === 'range') {
```
Consider making `range`, `radio` and `custom` enums or constants maybe?

jbocce commented on extensions/cornerstone-dicom-seg/src/panels/SegmentationToolbox.tsx:
```diff
@@ -0,0 +1,372 @@
+          name: 'Brush',
+          icon: 'icon-tool-brush',
+          disabled: !toolsEnabled,
+          active: state.activeTool === 'CircularBrush' || state.activeTool === 'SphereBrush',
```
Consider introducing constants for the following (note that this list is not exhaustive, so please check for more constants to add):

-'CircularBrush'
-'SphereBrush'
-'Brush'
-'Circle'
-'ThresholdCircularBrush'
- etc...

jbocce commented on modes/segmentation/.prettierrc:
```diff
@@ -0,0 +1,8 @@
+{
```
Do we need this file considering it is not consistent with that at the root?

jbocce commented on platform/core/src/services/HangingProtocolService/HangingProtocolService.ts:
```diff
@@ -1381,7 +1391,7 @@ export default class HangingProtocolService extends PubSubService {
     let highestSeriesMatchingScore = 0;
 
-    console.log('ProtocolEngine::matchImages', studyMatchingRules, seriesMatchingRules);
+    // console.log('ProtocolEngine::matchImages', studyMatchingRules, seriesMatchingRules);
```
Do we want to uncomment this?

jbocce commented on platform/ui/src/components/ButtonGroup/ButtonGroup.tsx:
```diff
@@ -1,150 +1,63 @@
 ButtonGroup.propTypes = {
-  children: PropTypes.node.isRequired,
+  buttons: PropTypes.arrayOf(PropTypes.object).isRequired,
+  orientation: PropTypes.oneOf(['horizontal', 'vertical']),
```
To avoid using strings, consider declaring this as you did below`PropTypes.oneOf(Object.values(ButtonEnums.orientation))`.

jbocce commented on platform/ui/src/utils/getMaxDigits.ts:
```diff
@@ -0,0 +1,7 @@
+const getMaxDigits = (maxValue: number, step: number) => {
```
Please add a comment for this function.

jbocce commented on extensions/cornerstone-dicom-seg/src/commandsModule.ts:
```diff
@@ -0,0 +1,378 @@
+
+      const { label } = segmentation;
+      const SeriesDescription =
+        promptResult.value !== undefined && promptResult.value !== ''
```
How about just `promptResult.value` ?

sedghi replied:
but we don't want ' '

jbocce replied:
The empty string ('') is falsy, so I think we are ok, no? https://developer.mozilla.org/en-US/docs/Glossary/Falsy

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.css:
```diff
@@ -0,0 +1,5 @@
+.input-range-thumb-design {
+  border: 2px solid #090c29;
+  border-radius: 50%;
+  background: #5acce6;
```
Can we use `@apply bg-primary-light` here instead?

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.css:
```diff
@@ -0,0 +1,5 @@
+.input-range-thumb-design {
+  border: 2px solid #090c29;
```
`@apply border-[2px] border-solid border-primary-dark`?

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.tsx:
```diff
@@ -0,0 +1,219 @@
+
+  useEffect(() => {
+    return () => {
+      window.removeEventListener('mousemove', handleMouseMove);
```
I might be wrong but I think we might have a slight memory leak here - ableit unlikely to occur. During a drag of either thumb, `InputDoubleRange` is rerendered constantly and as such a "new" `handleGlobalMouseUp` and `handleMouseMove` functions are created for each render. When these `removeEventListener` functions are called, the referenced `handleGlobalMouseUp` and `handleMouseMove` are NOT the ones that were added in `handleMouseDown`. However this would only be a problem if the component were unmounted during a drag. So maybe we don't need to worry about it too much, but I think we can solve the problem by simply using `useCallback` for each listner function.

sedghi replied:
Thanks for noticing this, but as you said let's worry when we have a bug for this

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.tsx:
```diff
@@ -0,0 +1,219 @@
+      <div
+        className="relative flex h-10 w-full items-center"
+        onMouseDown={handleMouseDown}
+        onMouseMove={handleMouseMove}
```
I don't think we need to add the `onMouseMove` and `onMouseUp` here since they are being added on the window.

sedghi replied:
No they should be there, they decide which thumb is dragged and move it

jbocce replied:
Ok. I find it odd that we need both these AND the listeners on the window, but perhaps I am missing something. In fact, I think maybe the mouse move listener on the window has a stale/null selected thumb and might be doing nothing. Maybe only the mouse up listener on the window is useful. Anyway please have a another look.

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.tsx:
```diff
@@ -0,0 +1,219 @@
+      updatedRangeValue[selectedThumb] = clampedValue;
+      setRangeValue(updatedRangeValue);
+
+      // No need to call onChange here, since we're updating local state+
```
I am confused by this comment...if there is no need to call onChange, then why are we doing it on the next line?

jbocce commented on platform/ui/src/components/InputRange/InputRange.css:
```diff
@@ -4,23 +4,32 @@ input[type='range'] {
 input[type='range']::-webkit-slider-thumb {
   -webkit-appearance: none;
-  border: none;
+  border: 2px solid #090c29;
```
Can we use @apply in here?

jbocce commented on platform/ui/src/components/InputRange/InputRange.tsx:
```diff
@@ -38,72 +42,76 @@ const InputRange: React.FC<{
+          id="myRange"
+          step={step}
+        />
+        <div style={{ width: labelWidth }}>
```
If `showLabel` is `false`, the `div` will be empty so do we still want to have this empty `div` occupy "width" space?

jbocce commented on platform/ui/src/components/PanelSection/PanelSection.tsx:
```diff
@@ -0,0 +1,54 @@
+import { Icon } from '@ohif/ui';
+import PropTypes from 'prop-types';
+
+const PanelSection = ({ title, children, actionIcons = [] }) => {
```
Add `children` and `actionIcons` to propTypes below.

jbocce commented on platform/ui/src/components/PanelSection/PanelSection.tsx:
```diff
@@ -0,0 +1,54 @@
+import PropTypes from 'prop-types';
+
+const PanelSection = ({ title, children, actionIcons = [] }) => {
+  const [isChildrenVisible, setChildrenVisible] = useState(true);
```
`areChildrenVisible` reads better in English, but it is not a big deal.

jbocce commented on platform/ui/src/components/SegmentationGroupTable/__stories__/segmentationGroupTable.stories.mdx:
```diff
@@ -55,7 +54,7 @@ SegmentationGroupTable is a component that renders the SegmentationGroupTables.
             null,
             {
-              label: 'Segment 1',
+              label: 'Segment 1 Segment 1Segment 1Segment 1Segment',
```
If we are trying to show how a long label is handled, then I would suggest changing this to something like 'This is a very long segment label'.

jbocce commented on platform/ui/src/components/SegmentationGroupTable/SegmentationGroupTable.tsx:
```diff
@@ -31,71 +41,123 @@ const SegmentationGroupTable = ({
+          activeSegmentation && [
+            {
+              name: 'settings-bars',
+              onClick: () => setIsConfigOpen(!isConfigOpen),
```
I think it is safest to use `setIsConfigOpen(isOpen => !isOpen)`

jbocce commented on platform/ui/src/components/SegmentationGroupTable/SegmentationGroupTable.tsx:
```diff
@@ -31,71 +41,123 @@ const SegmentationGroupTable = ({
+    }
+
+    setActiveSegmentationId(activeSegmentationIdToSet);
+  }, [segmentations, activeSegmentationId]);
```
I don't think `activeSegmentationId` should be a dependency for this `useEffect`.

sedghi commented on platform/ui/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -5,16 +5,48 @@ import classnames from 'classnames';
   const DropdownItem = useCallback(({ id, title, icon, onClick }) => {
+    // Split the title into lines of length maxCharactersPerLine
+    const lines = [];
+    for (let i = 0; i < title.length; i += maxCharactersPerLine) {
```
I mean what choice do we have, it is either a very wide dropdown as a result of only one of the items being longs or break like this. In practice I have not seen this happening

jbocce replied:
Well because it is an edge case we should leave it, but I suspect there would be something we could do with css https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_text/Wrapping_breaking_text

sedghi commented on modes/segmentation/src/initToolGroups.ts:
```diff
@@ -0,0 +1,206 @@
+      { toolName: toolNames.StackScrollMouseWheel, bindings: [] },
+    ],
+    passive: [
+      { toolName: toolNames.CircleScissors },
```
One is for deafult and one is for MPR

jbocce replied:
I understand, but I just thought maybe we could maybe introduce a constant and use the constant twice instead of the whole chunk of code. No big deal either way.

jbocce commented on platform/ui/src/components/SegmentationGroupTable/SegmentationGroupTable.tsx:
```diff
@@ -1,28 +1,38 @@
+  // segmentation initial config
+  segmentationConfig,
+  // UI show/hide
+  disableEditing,
```
Please ensure the propTypes below are up-to-date.

jbocce commented on platform/ui/src/components/SegmentationGroupTable/SegmentationGroupSegment.tsx:
```diff
@@ -21,200 +21,159 @@ const SegmentItem = ({
+        onMouseLeave={() => setIsNumberBoxHovering(false)}
       >
-        {isSegmentIndexHovering && showSegmentDelete ? (
+        {isNumberBoxHovering && showDelete ? (
```
Not a big deal, but I believe instead of using `onMouseEnter` and `onMouseLeave` to keep track of the hover we could also add a `group` class to the parent `div` and then have class names `hidden group-hover:block` on the `Icon`.

jbocce commented on platform/ui/src/components/SegmentationGroupTable/SegmentationGroupSegment.tsx:
```diff
@@ -21,200 +21,159 @@ const SegmentItem = ({
+            {}
+          )}
+        >
+          {!isVisible && !isRowHovering && (
```
It is a bit confusing that each icon is rendered here when NOT hovering and then rendered on line 131 when hovering. It would be great if we only had to "look" in one place for each icon. But if it is too much trouble, then that is ok.

sedghi replied:
seems like there is another bug

![CleanShot 2023-09-14 at 13 10 34](https://github.com/OHIF/Viewers/assets/7490180/ee6212c7-09ef-44c3-be20-ae578f3aa1e0)

jbocce commented on platform/ui/src/components/SegmentationGroupTable/SegmentationDropDownRow.tsx:
```diff
@@ -0,0 +1,138 @@
+import { Select, Icon, Dropdown } from '../../components';
+import PropTypes from 'prop-types';
+
+function SegmentationDropDownRow({
```
Update propTypes below.

jbocce reviewed (changes requested):
It is looking good. Lots of 'neat' GUI controls.

Have a look at my comments. I will now do some more testing.

jbocce commented:
I believe the former issues are handled for the most part. I think for item 17, perhaps the reconstruction for some series/studies is just slow with Orthanc and the gray box appears for a "while". Here are some more things I found.

a. After hydrating a saved/reported segmentation, the thumbnails in the side study browser go blank.
b. The stack sync tool appears to be broken. For example, when navigating one of two axial series on screen and the tool is activated only one scrolls. Actually, it appears that the even though I click the tool to activate it, it appears not selected/active.
c. After deleting the last segmentation, the brush tools are still active and using them throws exceptions in the console.
d. After deleting all segments, the last tool is still active and it can be used.
e. I noticed that setting the fill opacity to 0 makes the outline disappear (i.e. I think it goes transparent). Is this expected? I would have expected the outline to remain but with no fill.
f. So I still see the issue where the window level appears to change when adding a segmentation (i.e. item 19 from before) and after having a good look at the images and the various stack scroll bars/sliders) the middle slice is NOT being moved to from what I can tell. If anything I would say the window level of the middle image is being applied yet the current/last image remains on screen.

https://github.com/OHIF/Viewers/assets/109477394/ac247769-0b8b-405a-89d8-84f8041dd56c

jbocce commented:
The stack sync tool appears to work now. However, once a segmentation is added/drawn, the scroll syncing stops. Is this expected because now we have a volume viewport?

jbocce commented:
I still see the problem where after deleting the last segmentation, the brush tools are still active and using them throws exceptions in the console.

jbocce commented:
I slide the fill opacity all the way to the left and the label says 4%. Is this expected?

![image](https://github.com/OHIF/Viewers/assets/109477394/2fe61d5c-ef03-42bf-9514-919b59502624)

jbocce commented:
I am still seeing some weirdness with the threshold tool slider. Consider the following steps...

1. Drag the far left thumb of the slider up and over to the left. Drag the mouse over the word Mode in the panel above.
2. Notice it is selecting/highlighting text during the drag this is not desirable.
3. Release the drag when over the word Mode. Notice that text is still highlighted.
4. Try to drag the left thumb again. Notice the weird dragging of text that is happening.
5. Release the mouse button and continue moving the mouse and notice that we are "stuck" in drag mode.

https://github.com/OHIF/Viewers/assets/109477394/0b116551-5251-4c11-a156-cf3b207f6154

jbocce commented:
After saving a seg report, if the newly saved series/report is double clicked. The following exception is thrown in the console...

Uncaught TypeError: segDisplaySet.instance is not iterable (cannot read property Symbol(Symbol.iterator))
    at _getReferencedDisplaySetMetadata (C:\dev\Viewers\extensions\cornerstone-dicom-seg\src\viewports\OHIFCornerstoneSEGViewport.tsx:379:1)
    at OHIFCornerstoneSEGViewport (C:\dev\Viewers\extensions\cornerstone-dicom-seg\src\viewports\OHIFCornerstoneSEGViewport.tsx:68:1)
    at renderWithHooks (C:\dev\Viewers\node_modules\react-dom\cjs\react-dom.development.js:14985:1)
    at updateFunctionComponent (C:\dev\Viewers\node_modules\react-dom\cjs\react-dom.development.js:17356:1)
    at mountLazyComponent (C:\dev\Viewers\node_modules\react-dom\cjs\react-dom.development.js:17677:1)
    at beginWork (C:\dev\Viewers\node_modules\react-dom\cjs\react-dom.development.js:19055:1)
    at HTMLUnknownElement.callCallback (C:\dev\Viewers\node_modules\react-dom\cjs\react-dom.development.js:3945:1)
    at Object.invokeGuardedCallbackDev (C:\dev\Viewers\node_modules\react-dom\cjs\react-dom.development.js:3994:1)
    at invokeGuardedCallback (C:\dev\Viewers\node_modules\react-dom\cjs\react-dom.development.js:4056:1)
    at beginWork$1 (C:\dev\Viewers\node_modules\react-dom\cjs\react-dom.development.js:23964:1)
    
 

https://github.com/OHIF/Viewers/assets/109477394/059ce55b-5300-4f57-99f9-0e55ec6cf12e

jbocce commented:
Consider removing the rectangle tool for segmentation mode.

jbocce commented:
Deleting a hyrdrated seg should remove the corresponding highlight in the study list/thumbnail.

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.tsx:
```diff
@@ -0,0 +1,225 @@
+      <div
+        className="relative flex h-10 w-full items-center"
+        onMouseDown={handleMouseDown}
+        // onMouseMove={handleMouseMove}
```
Remove if not needed.

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.tsx:
```diff
@@ -0,0 +1,225 @@
+  };
+
+  const handleMouseMove = e => {
+    console.debug('🚀 ~ selectedThumb:', selectedThumbRef.current);
```
I don't think we want this during a drag.

jbocce reviewed (changes requested):
Just a few issues to address.

sedghi commented:
> The stack sync tool appears to work now. However, once a segmentation is added/drawn, the scroll syncing stops. Is this expected because now we have a volume viewport?

Yes it is because it is converted to volume viewport, ideally we should have viewportSync instead of `stack` viewport sync which handles both stack and volume viewport

> I still see the problem where after deleting the last segmentation, the brush tools are still active and using them throws exceptions in the console.

Probably not built correctly?

https://github.com/OHIF/Viewers/assets/7490180/47bfe02d-743d-433e-b3d1-2844a5c1ec17

> I slide the fill opacity all the way to the left and the label says 4%. Is this expected?

We don't have proper 0 right now as it removes the outline as we discussed, so I just put it as min 4 for now

>

jbocce commented:
> I still see the problem where after deleting the last segmentation, the brush tools are still active and using them throws exceptions in the console.

This looks fixed now - thanks. However, the window level tool appears to be active, yet there is no indication that it is active. (i.e. the tool  button does not indicate that it is selected).

jbocce commented:
It appears that adding segments to a different oriented series is NOT support or at least throws an exceptions. Is there a better way to handle this? Perhaps it is something we address later, but I figured it would be worth mentioning here at least in case there was something that we could easily do (perhaps disable the tools or something).

```
Uncaught Error: Only source data the same dimensions/size/orientation as the segmentation currently supported.
    at o.Ka (fillSphere.ts:117:11)
    at o.applyActiveStrategy (BaseTool.ts:85:39)
    at Rr._endCallback (BrushTool.ts:383:10)
    at se (triggerEvent.ts:37:13)
    at HTMLDocument.ct (mouseDownListener.ts:327:5)
```
https://github.com/OHIF/Viewers/assets/109477394/0fa50225-d96a-410c-b66e-c20ce6481ebd

jbocce commented:
Looks like there is a failing cypress test. Please have a look.

jbocce commented:
I noticed that there is an exception thrown when saving measurements that does not appear to exist in master. However, there appears to be an exception thrown when hydrating measurements in master. I will log a bug for that.

jbocce commented:
Hmm. Not sure why I didn't see this earlier, but when deleting a segmentation I am getting the following exception 

Perhaps it is only seen when displaying multiple series/orientations???

```
Uncaught TypeError: viewport.removeVolumeActors is not a function
    at removeLabelmapFromElement (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\tools\displayTools\Labelmap\removeLabelmapFromElement.ts:22:1)
    at _removeLabelmapFromToolGroupViewports (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\tools\displayTools\Labelmap\labelmapDisplay.ts:485:1)
    at Object.removeSegmentationRepresentation (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\tools\displayTools\Labelmap\labelmapDisplay.ts:116:1)
    at _removeSegmentation (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\stateManagement\segmentation\removeSegmentationsFromToolGroup.ts:74:1)
    at C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\stateManagement\segmentation\removeSegmentationsFromToolGroup.ts:57:1
    at Array.forEach (<anonymous>)
    at Module.removeSegmentationsFromToolGroup (C:\dev\Viewers\node_modules\@cornerstonejs\tools\src\stateManagement\segmentation\removeSegmentationsFromToolGroup.ts:56:1)
    at C:\dev\Viewers\extensions\cornerstone\src\services\SegmentationService\SegmentationService.ts:1993:1
    at Array.forEach (<anonymous>)
    at SegmentationService._removeSegmentationFromCornerstone (C:\dev\Viewers\extensions\cornerstone\src\services\SegmentationService\SegmentationService.ts:1981:1)
```


https://github.com/OHIF/Viewers/assets/109477394/705d24e9-f9f4-4152-a322-dfb53acd216c

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.tsx:
```diff
@@ -0,0 +1,223 @@
+      <div
+        className="relative flex h-10 w-full items-center"
+        onMouseDown={handleMouseDown}
+        onMouseUp={handleMouseUp}
```
It is great to see that we don't need the onMouseMove here any more. Do we still need the onMouseUp?

jbocce commented on platform/core/src/services/ViewportGridService/ViewportGridService.ts:
```diff
@@ -74,6 +74,11 @@ class ViewportGridService extends PubSubService {
     return this.serviceImplementation._getState();
   }
 
+  public getActiveViewportId() {
```
Please add this to the docs if you haven't already done so. Thanks.

jbocce commented on extensions/cornerstone-dicom-seg/src/panels/SegmentationToolbox.tsx:
```diff
@@ -0,0 +1,404 @@
+            },
+            {
+              type: 'custom',
+              id: 'threshold-range',
```
Consider using `'segmentation-threshold-range'` to guarantee better uniqueness. I have seen browsers do crazy things when ids are no unique.

jbocce reviewed (changes requested):
Just a few issues to address.

jbocce commented:
The following exception is thrown when attempting to relaunch Segmentation mode after adding a segmentation in the last session.

```
ToolGroupService.ts:166 Uncaught (in promise) Error: ToolGroup default already exists
    at ot.createToolGroup (ToolGroupService.ts:166:13)
    at ot.createToolGroupAndAddTools (ToolGroupService.ts:188:28)
    at initToolGroups.ts:53:20
    at g (initToolGroups.ts:78:3)
    at Object.onModeEnter (index.tsx:58:7)
    at Mode.tsx:334:13
    at Mode.tsx:394:5
    at Ul (react-dom.production.min.js:262:359)
    at t.unstable_runWithPriority (scheduler.production.min.js:18:343)
    at qa (react-dom.production.min.js:122:325)
```


https://github.com/OHIF/Viewers/assets/109477394/f316e048-02fa-481d-bec5-4e62c5aa6c60

jbocce commented on platform/ui/src/components/InputDoubleRange/InputDoubleRange.tsx:
```diff
@@ -0,0 +1,222 @@
+    };
+  }, []);
+
+  const handleMouseUp = () => {
```
This I think is no longer referenced so it can be removed.

jbocce commented on platform/docs/docs/platform/services/ui/viewport-grid-service.md:
```diff
@@ -32,6 +32,7 @@ is expected to support, [check out it's interface in `@ohif/core`][interface]
 | `reset()`                                                             | Resets the default states                           |
 | `getNumViewportPanes()`                                               | Gets the number of visible viewport panes           |
 | `getLayoutOptionsFromState(gridState)`                                | Utility method that produces a `ViewportLayoutOptions` based on the passed in state|
+| `getActiveViewportId()`                                | Returns the viewportId of the active viewport in the grid|
```
Add a space between viewport and id.

jbocce reviewed (changes requested):
Just a couple more things.

jbocce commented:
Choosing a tool/item from either split button menu (i.e. measurement or more tools) no longer shows the tool as selected. Also the mpr button does not render as selected when in mpr mode.

https://github.com/OHIF/Viewers/assets/109477394/9e36aa15-4073-4df0-9cde-8bb49f2f82db

jbocce commented on platform/core/src/services/ToolBarService/ToolbarService.ts:
```diff
@@ -94,8 +94,9 @@ export default class ToolbarService extends PubSubService {
-          // only set the primary tool if no error was thrown
-          this.state.primaryToolId = itemId;
+          // only set the primary tool if no error was thrown and if the itemId is not
+          // undefined set the first tool as the primary tool
```
It is not clear from the comment why we are setting the primaryToolId the way we are. Why might `itemId` be undefined? Also the comment suggests that the first tool is used if `itemId` is not undefined.

sedghi replied:
for backward compatibility i put it like this, we don't really need itemId

jbocce commented on extensions/default/src/Toolbar/ToolbarButtonWithServices.tsx:
```diff
@@ -0,0 +1,75 @@
+import React, { useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+
+function ToolbarSplitButtonWithServices({
```
Shouldn't this be `ToolbarButtonWithServices`?

jbocce commented:
I am still seeing some weirdness with the toolbar buttons...

When stack image sync is active, I sometimes see this... 
![image](https://github.com/OHIF/Viewers/assets/109477394/ba75dff6-0f12-4a99-ba6c-ec7f9566b7ab)
and sometimes this...
![image](https://github.com/OHIF/Viewers/assets/109477394/0cee0dc5-9b58-41ad-a55b-cc954622fde7)

jbocce commented on extensions/default/src/Toolbar/ToolbarSplitButtonWithServices.tsx:
```diff
@@ -0,0 +1,192 @@
+    items: getSplitButtonItems(items).filter(item =>
+      isRadio && !isAction ? item.id !== primary.id : true
+    ),
+    isHovering: false,
```
Are `isHovering` and `isExpanded` actually used somewhere?

jbocce commented on extensions/default/src/Toolbar/ToolbarSplitButtonWithServices.tsx:
```diff
@@ -0,0 +1,192 @@
+ToolbarSplitButtonWithServices.propTypes = {
+  isRadio: PropTypes.bool,
+  isAction: PropTypes.bool,
+  bState: PropTypes.shape({
```
Please remove and/or update this list such that it is accurate.

jbocce reviewed (approved):
Some minor comments, but otherwise approved.

Thanks for the tremendous effort on this.
